api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto

```diff
@@ -17,10 +20,42 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 //
 // [#extension: envoy.http.original_ip_detection.xff]
 message XffConfig {
-  // The number of additional ingress proxy hops from the right side of the
-  // :ref:`config_http_conn_man_headers_x-forwarded-for` HTTP header to trust when
-  // determining the origin client's IP address. The default is zero if this option
-  // is not specified. See the documentation for
-  // :ref:`config_http_conn_man_headers_x-forwarded-for` for more information.
-  uint32 xff_num_trusted_hops = 1;
+  oneof xff_trusted_config {
+    option (validate.required) = true;
+
+    // The number of additional ingress proxy hops from the right side of the
+    // :ref:`config_http_conn_man_headers_x-forwarded-for` HTTP header to trust when
+    // determining the origin client's IP address. The default is zero if this option
+    // is not specified. See the documentation for
+    // :ref:`config_http_conn_man_headers_x-forwarded-for` for more information.
+    uint32 xff_num_trusted_hops = 1;
+
+    // The `CIDR <https://tools.ietf.org/html/rfc4632>`_ ranges to trust when
+    // evaluating the remote IP address to determine the origin client's IP address.
+    // This is used instead of
+    // :ref:`use_remote_address <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.use_remote_address>`.
+    // When the remote IP address matches a trusted CIDR and the
+    // :ref:`config_http_conn_man_headers_x-forwarded-for` header was sent, the last
+    // IP address from ``x-forwarded-for`` is used as the original client address.
+    // If no ``x-forwarded-for`` header was sent, the remote IP is considered the real IP.
+    //
+    // If ``recurse`` is set the :ref:`config_http_conn_man_headers_x-forwarded-for`
+    // header is also evaluated against the trusted CIDR list, and the last non-trusted
+    // address is used as the original client address.
+    //
+    // This is typically used when requests are proxied by a
+    // `CDN <https://en.wikipedia.org/wiki/Content_delivery_network>`_.
+    XffTrustedCidrs xff_trusted_cidrs = 2;
+  }
+}
+
+message XffTrustedCidrs {
+  // The list of `CIDRs <https://tools.ietf.org/html/rfc4632>`_ from which remote
+  // connections are considered trusted.
+  repeated config.core.v3.CidrRange cidrs = 1;
+
+  // If ``true``, recurse through the :ref:`config_http_conn_man_headers_x-forwarded-for`
+  // HTTP header to find the first (from the right) non-trusted IP address; otherwise
+  // the last address in ``x-forwarded-for`` is used.
+  bool recurse = 2;
```

Comment:
Will all data planes always want false to be the default?  If not, consider using `google.protobuf.BoolValue` here, so that the control plane can differentiate between "explicitly set to false" and "unset, use default value".

Reply:
Updated.